src/goto-instrument/nondet_volatile.cpp

```diff
@@ -101,25 +121,102 @@ nondet_volatile(symbol_tablet &symbol_table, goto_programt &goto_program)
+      nondet_volatile_lhs(symbol_table, code_function_call.lhs(), should_havoc);
     }
     else if(instruction.has_condition())
     {
       // do condition
       exprt cond = instruction.get_condition();
-      nondet_volatile_rhs(symbol_table, cond);
+      nondet_volatile_rhs(symbol_table, cond, should_havoc);
       instruction.set_condition(cond);
     }
   }
 }
 
-void nondet_volatile(goto_modelt &goto_model)
+void nondet_volatile(goto_modelt &goto_model, havoc_predt should_havoc)
 {
   Forall_goto_functions(f_it, goto_model.goto_functions)
-    nondet_volatile(goto_model.symbol_table, f_it->second.body);
+    nondet_volatile(goto_model.symbol_table, f_it->second.body, should_havoc);
 
   goto_model.goto_functions.update();
 }
+
+void parse_nondet_volatile_options(const cmdlinet &cmdline, optionst &options)
+{
+  PRECONDITION(!options.is_set(NONDET_VOLATILE_OPT));
+  PRECONDITION(!options.is_set(NONDET_VOLATILE_VARIABLE_OPT));
+
+  const bool nondet_volatile_opt = cmdline.isset(NONDET_VOLATILE_OPT);
+  const bool nondet_volatile_variable_opt
+    = cmdline.isset(NONDET_VOLATILE_VARIABLE_OPT);
+
+  if(nondet_volatile_opt && nondet_volatile_variable_opt)
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "only one of " NONDET_VOLATILE_OPT "/" NONDET_VOLATILE_VARIABLE_OPT "can "
+      "be given at a time",
+      NONDET_VOLATILE_OPT "/" NONDET_VOLATILE_VARIABLE_OPT);
+  }
+
+  if(nondet_volatile_opt)
+  {
+    options.set_option(NONDET_VOLATILE_OPT, true);
+  }
+  else if(cmdline.isset(NONDET_VOLATILE_VARIABLE_OPT))
+  {
+    options.set_option(
+      NONDET_VOLATILE_VARIABLE_OPT,
+      cmdline.get_values(NONDET_VOLATILE_VARIABLE_OPT));
+  }
+}
+
+void nondet_volatile(goto_modelt &goto_model, const optionst &options)
+{
+  if(options.get_bool_option(NONDET_VOLATILE_OPT))
+  {
+    nondet_volatile(goto_model);
+  }
+  else if(options.is_set(NONDET_VOLATILE_VARIABLE_OPT))
+  {
+    const auto &variable_list =
+      options.get_list_option(NONDET_VOLATILE_VARIABLE_OPT);
+
+    std::set<irep_idt> variables(variable_list.begin(), variable_list.end());
+    const namespacet ns(goto_model.symbol_table);
+
+    // typecheck given variables
+    for(const auto &id : variables)
+    {
+      const symbolt *symbol;
+
+      if(ns.lookup(id, symbol))
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "given name " + id2string(id) + " not found in symbol table",
+          NONDET_VOLATILE_VARIABLE_OPT);
+      }
+
+      if(symbol->is_type || !symbol->type.get_bool(ID_C_volatile))
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "given name " + id2string(id) + " does not represent a volatile "
+          "variable",
+          NONDET_VOLATILE_VARIABLE_OPT);
+      }
+
+      INVARIANT(
+        symbol->type.id() != ID_code,
+        "symbol must not represent a function");
+    }
+
+    auto should_havoc = [&variables](const exprt &expr) {
```

Comment:
:no_entry_sign: This looks like a dangling reference..? You capture variables by reference, but they are a local variable that goes out of scope, but you return the lambda. I think `=variables` is needed here, or convince me why I'm wrong!

Reply:
The lambda is not returned. `variables`, the lambda definition, and the call to `nondet_volatile()` (which uses the lambda) are all in the same scope.

Reply:
Hmm - must have been half asleep - right you are.